src/gurt/telemetry.c

```diff
@@ -2513,6 +2554,224 @@ get_unique_shmem_key(const char *path, int id)
+	if (unlikely(rc != 0)) {
+		D_ERROR("failed to get producer mutex\n");
+		D_GOTO(out_ephemeral_unlock, rc);
+	}
+
+	link_node = d_tm_find_metric(ctx, path);
+	if (link_node == NULL) {
+		D_ERROR("nonexistent metric: %s", path);
+		D_GOTO(out_unlock, rc = -DER_NONEXIST);
+	}
+
+	rc = shm_stat_key(link_node->dtn_shmem_key, &shminfo, NULL);
+	if (rc < 0) {
+		D_ERROR("failed to stat parent segment\n");
+		D_GOTO(out_unlock, rc = -DER_MISC);
+	}
+	o_uid = shminfo.shm_perm.uid;
+
+	rc = shm_stat_key(child_key, &shminfo, &child_shmid);
+	if (rc < 0) {
+		D_ERROR("failed to stat child segment\n");
+		D_GOTO(out_unlock, rc = -DER_MISC);
+	}
+
+	if (o_uid == shminfo.shm_perm.uid)
+		D_GOTO(out_unlock, rc = 0);
+
+	shminfo.shm_perm.uid = o_uid;
+	rc                   = shmctl(child_shmid, IPC_SET, &shminfo);
+	if (rc != 0) {
+		D_ERROR("failed to set child segment ownership: %d", rc);
+		rc = -DER_MISC;
+	}
+
+out_unlock:
+	d_tm_unlock_shmem();
+out_ephemeral_unlock:
+	if (tm_shmem.multiple_writer_lock)
+		D_MUTEX_UNLOCK(&ctx->shmem_root->sh_multiple_writer_lock);
+
+	return rc;
+}
+
+static int
+attach_path_segment(key_t key, char *path)
+{
+	struct d_tm_node_t       *link_node;
+	struct d_tm_context      *ctx = tm_shmem.ctx;
+	struct d_tm_shmem_hdr    *parent_shmem;
+	struct d_tm_metric_t     *link_metric;
+	struct shmem_region_list *region_entry;
+	int                       rc;
+
+	if (unlikely(path == NULL)) {
+		D_ERROR("NULL inputs\n");
+		D_GOTO(fail, rc = -DER_INVAL);
+	}
+
+	if (tm_shmem.multiple_writer_lock)
+		D_MUTEX_LOCK(&ctx->shmem_root->sh_multiple_writer_lock);
+
+	rc = d_tm_lock_shmem();
+	if (unlikely(rc != 0)) {
+		D_ERROR("failed to get producer mutex\n");
+		D_GOTO(fail_ephemeral_unlock, rc);
+	}
+
+	link_node = d_tm_find_metric(ctx, path);
+	if (link_node != NULL) {
+		D_GOTO(fail_unlock, rc = -DER_EXIST);
+	}
+
+	/* Add a link to the new region */
+	rc = add_metric(ctx, &link_node, D_TM_LINK, NULL, NULL, path);
+	if (rc != 0) {
+		D_ERROR("can't set up the link node, " DF_RC "\n", DP_RC(rc));
+		D_GOTO(fail_tracking, rc);
+	}
+
+	/* track attached regions within the parent shmem */
+	parent_shmem = get_shmem_for_key(ctx, link_node->dtn_shmem_key);
+	if (parent_shmem == NULL) {
+		D_ERROR("failed to get parent shmem pointer\n");
+		D_GOTO(fail_link, rc = -DER_NO_SHMEM);
+	}
+
+	D_ASSERT(link_node->dtn_type == D_TM_LINK);
+	link_metric                 = conv_ptr(parent_shmem, link_node->dtn_metric);
+	link_metric->dtm_data.value = key;
+
+	rc = get_free_region_entry(parent_shmem, &region_entry);
+	if (rc != 0)
+		D_GOTO(fail_link, rc);
+	region_entry->rl_key       = key;
+	region_entry->rl_link_node = link_node;
+
+	d_tm_unlock_shmem();
+	if (tm_shmem.multiple_writer_lock)
+		D_MUTEX_UNLOCK(&ctx->shmem_root->sh_multiple_writer_lock);
+	return 0;
+
+fail_link:
+	invalidate_link_node(parent_shmem, link_node);
+fail_tracking:
+	close_shmem_for_key(ctx, key, true);
+	goto fail_unlock; /* shmem will be closed/destroyed already */
```

Comment:
A remnant from an earlier implementation?

Reply:
Yeah, this isn't needed here.